functionary/core/api/v1/utils.py

```diff
@@ -0,0 +1,114 @@
+import json
+import re
+from collections import OrderedDict
+from typing import Union
+
+from django.core.files.uploadedfile import InMemoryUploadedFile
+from rest_framework import serializers
+from rest_framework.exceptions import ValidationError
+
+from core.models import Function, FunctionParameter
+from core.utils.parameter import PARAMETER_TYPE
+
+PREFIX = "param"
+SEPARATOR = r"\."
+
+r = re.compile(rf"^({PREFIX}){SEPARATOR}(\w+)$")
+
+
+def parse_parameters(values: OrderedDict) -> None:
+    """Mutate given values to move all `param.` arguments in parameters field"""
+    _update_parameter_field_type(values)
+
+    # Avoid dictionary changed size error by wrapping items in list
+    for param, _ in list(values.items()):
+        if not (valid_param := get_parameter_name(param)):
+            continue
+
+        value = values.pop(param)
+        if type(value) == InMemoryUploadedFile:
+            value = value.name
+
+        values["parameters"][valid_param.group(2)] = value
+
+
+def cast_parameters(values: OrderedDict) -> None:
+    """Parent method to cast parameters to necessary types"""
+    _cast_json_parameters(values)
+
+
+def get_parameter_name(parameter: str) -> Union[re.Match, None]:
+    """Return match if parameter name is of valid format"""
+    return r.match(parameter)
+
+
+def _cast_json_parameters(values: OrderedDict) -> None:
+    """Mutates all JSON parameter fields into python objects
+
+    Args:
+        values: An OrderedDict containing the values passed to the serializer
+
+    Returns:
+        None
+
+    Raises:
+        ValidationError: When a JSON parameter is not valid JSON
+    """
+    function = _get_function(values)
+
+    json_params: list[FunctionParameter] = [
+        param
+        for param in function.parameters.all()
+        if param.parameter_type == PARAMETER_TYPE.JSON
+    ]
+
+    for json_param in json_params:
+        try:
+            if type(values["parameters"][json_param.name]) == dict:
+                continue
+
+            values["parameters"][json_param.name] = json.loads(
+                values["parameters"][json_param.name]
+            )
+        except json.decoder.JSONDecodeError as err:
+            exception_map = {json_param.name: err.msg}
+            exc = ValidationError(exception_map)
+            raise serializers.ValidationError(serializers.as_serializer_error(exc))
+
+
+def _get_function(values: OrderedDict) -> Function:
```

Comment:
Fixed. I opted to place the function validation after the `super().to_internal_value()` since I only wanted to access  the `function_name` and `package_name` fields once, and the `super().to_internal_value()` check expects those those two fields to exist. So I figured I could just pop those fields after the `super().to_internal_value()` check